### PR TITLE
Add support for ID-based links

### PR DESCRIPTION
**Pull request by wgslr**
Implement functionality inspired by [Zettlr](https://github.com/Zettlr/Zettlr) -- the ability to store ids in files and link to them with wiki-style links. For example:
```
file1.md

# This is a file having an id
This is its id: 20200522225822
```

```
file2.md

# This is a file linking to another
See the other file: [[20200522225822]]
```

The file id is the first match of a (configurable) regex in its contents.

### CODE REVIEW

File: src/extension.ts [hunk 1]
```diff
@@ -86,25 +107,52 @@ const parseFile = async (source: string) => {
...
   if (index !== -1) {
     nodes[index].label = title;
   } else {
-    nodes.push({ id: id(source), path: source, label: title });
+    nodes.push({ id: id(filePath), path: filePath, label: title });
   }
 
-  edges = edges.filter((edge) => edge.source !== id(source));
+  // remove edges based on an old version of this file
+  edges = edges.filter((edge) => edge.source !== filePath);
```
Comment (tchayen) [inline src/extension.ts:118]: Is it correct with `filePath` instead of `id(filePath)`?

File: src/extension.ts [hunk 2]
```diff
@@ -86,25 +107,52 @@ const parseFile = async (source: string) => {
...
+
   const links = findLinks(ast);
 
   for (const link of links) {
-    const parentDirectory = source.split("/").slice(0, -1).join("/");
-    const target = path.normalize(`${parentDirectory}/${link}`);
+    let target = link;
+    if (!path.isAbsolute(link)) {
+      const parentDirectory = filePath.split("/").slice(0, -1).join("/");
```
Comment (tchayen) [inline src/extension.ts:125]: Now when I look at it, this line might be moved out of the `for` loop, right?

File: src/extension.ts [hunk 3]
```diff
@@ -86,25 +107,52 @@ const parseFile = async (source: string) => {
...
+const findFileId = async (filePath: string): Promise<string | null> => {
+  const buffer = await vscode.workspace.fs.readFile(vscode.Uri.file(filePath));
+  const content = new TextDecoder("utf-8").decode(buffer);
+
+  const match = content.match(ID_REGEX);
+  if (match) {
+    return match[1];
+  } else {
+    return null;
+  }
```
Comment (tchayen) [inline src/extension.ts:142]:
```suggestion
  return match ? match[1] : null;
```

File: src/extension.ts [hunk 4]
```diff
@@ -69,12 +77,25 @@ const findTitle = (ast: MarkdownNode): string | null => {
   return null;
 };
 
-const parseFile = async (source: string) => {
-  const buffer = await vscode.workspace.fs.readFile(vscode.Uri.file(source));
+const idResolver = (id: string): string[] => {
```
Comment (tchayen) [inline src/extension.ts:80]:
```suggestion
const idResolver = (id: string) => {
```

I think type of this function will be correctly inferred by the TSC.

File: src/extension.ts [hunk 5]
```diff
@@ -69,12 +77,25 @@ const findTitle = (ast: MarkdownNode): string | null => {
...
+  const filePath = idToPath[id];
+  if (filePath === undefined) {
+    return [id];
+  } else {
+    return [filePath];
+  }
+};
+const parser = unified()
```
Comment (tchayen) [inline src/extension.ts:88]:
```suggestion
};

const parser = unified()
```

File: src/extension.ts (hunk 2 above)
Comment (wgslr) [reply src/extension.ts:125]: Correct. This is what happens when I push the results of `git merge` without giving them more attention

File: src/extension.ts (hunk 1 above)
Comment (wgslr) [reply src/extension.ts:118]: No :sweat_smile: Damn merges!

File: src/extension.ts (hunk 4 above)
Comment (wgslr) [reply src/extension.ts:80]: I like having explicit types whenever I have the patience to write them, but will adjust to your style

File: src/extension.ts (hunk 2 above)
Comment (wgslr) [reply src/extension.ts:126]: Oh, nevermind, that's not what you were talking about. Anyways, yes, I will move it.

File: README.md [hunk 6]
```diff
@@ -18,10 +18,39 @@ The graph refreshes automatically every time you:
 ## Concepts
 
 - Title is always the first markdown heading of depth 1, i.e. `# Title`.
-- Any link or reference to a local file causes the extension to attempt parsing it. The title is required but links are optional (it results in a lone node with no edges in such case).
-- Graph is not directed. It doesn't show which file has the link and which one is linked.
+- Files which do not have a title do not appear in the graph
+- Files can link to other files usin path-based or ID-based links
```
Comment (tchayen) [inline README.md:22]:
```suggestion
- Files which do not have a title do not appear in the graph.
- Files can link to other files using local Markdown links or ID-based links.
```

File: README.md [hunk 7]
```diff
@@ -18,10 +18,39 @@ The graph refreshes automatically every time you:
...
+- The graph is not directed. It doesn't show which file has the link and which one is linked.
 - Directory structure is not relevant for the graph. All that matters is the mutual links between files.
 
+### Path-based links
```
Comment (tchayen) [inline README.md:26]:
```suggestion
### Local Markdown links
```

File: src/extension.ts [hunk 8]
```diff
@@ -18,25 +18,45 @@ type Node = {
...
+const getConfiguration = (key: string) =>
+  vscode.workspace.getConfiguration("markdown-links")[key];
+
+const getFileIdRegexp = () => {
+  const DEFAULT_VALUE = "\\d{14}";
+  const userValue = getConfiguration("fileIdRegexp") || DEFAULT_VALUE;
+
+  // ensure the id is not preceeded by [[, which would make it a part of
+  // wiki-style link, and put the user-supplied regex in a capturing group to
+  // retrieve matching string
```
Comment (tchayen) [inline src/extension.ts:47]:
```suggestion
  // Ensure the id is not preceeded by [[, which would make it a part of
  // wiki-style link, and put the user-supplied regex in a capturing group to
  // retrieve matching string.
```

**Pull request merged**